python/ray/autoscaler/node_provider.py

```diff
@@ -276,3 +278,92 @@ def get_command_runner(self,
             return DockerCommandRunner(docker_config, **common_args)
         else:
             return SSHCommandRunner(**common_args)
+
+class NodeProviderCache:
+    """Base class for storing cached node information, used to avoid excessive
+    API calls to cloud providers in some cases.
+
+    This implementation stores all information in memory. 
+    """
+
+    def __init__(self):
+        # tag_map is supposed to be the superset of node_map.
+        self.node_map = {}
+        self.tag_map = {}
+        self.lock = threading.Lock()
+
+    def get_node(self, node_id):
+        with lock:
+            return self.node_map.get(node_id)
+
+    def set_node(self, node_id, node):
+        with lock:
+            self.node_map[node_id] = node
+            self.tag_map.setdefault(node_id, {})
+
+    def node_exists(self, node_id):
+        with lock:
+            return node_id in self.node_map
+
+    def get_tags(self, node_id):
+        with lock:
+            return self.tag_map.get(node_id)
+
+    def set_tags(self, node_id, tags):
+        with lock:
+            self.tag_map[node_id] = tags
+
+    def update_tags(self, node_id, tags):
+        with lock:
+            self.tag_map[node_id].update(tags)
+
+    def tags_exists(self, node_id):
+        with lock:
+            return node_id in self.tags_map
```

Comment:
This is inconsistent with the usage in  the `aws_node_provider`. It is used as `tags_exist` (singular exist). Can we rename both to `tag_exists` [singular tag, plural exist]

Reply:
okay.